### ISSUE TITLE
Flutter app with Sentry calling a swift package with sentry leads to two sets of classes and causes undefined behaviour and crashes 

### ISSUE THREAD

**Issue by julianharty** [getsentry/sentry-dart#1618]
### Platform

Dart

### Obfuscation

Enabled

### Debug Info

Enabled

### Doctor

```
% flutter doctor -v                                                                    
[✓] Flutter (Channel stable, 3.13.1, on macOS 13.4.1 22F82 darwin-arm64, locale en-GB)
    • Flutter version 3.13.1 on channel stable at /opt/homebrew/Caskroom/flutter/3.13.1/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e1e47221e8 (7 days ago), 2023-08-22 21:43:18 -0700
    • Engine revision b20183e040
    • Dart version 3.1.0
    • DevTools version 2.25.0

[!] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/julianharty/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.81.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.70.0

[✓] Connected device (3 available)
    • Julian's iPhone (mobile) • 00008110-000165D43A08401E • ios            • iOS 16.6 20G75
    • macOS (desktop)          • macos                     • darwin-arm64   • macOS 13.4.1 22F82 darwin-arm64
    • Chrome (web)             • chrome                    • web-javascript • Google Chrome 116.0.5845.110

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
Note: the Android licenses have been accepted; I doubt their status is relevant to the issue I'm reporting. `flutter doctor --android-licenses` fails to run in this terminal window because it doesn't like the java version (19.0.1)

### Version

8.9.6 in Swift; 7.6.3 in Flutter

### Context
1. We have an SDK written in Swift that uses Sentry via Swift Package Manager.
2. A client have an app written in Flutter and they want to use our Swift package. They also use Sentry for error reporting.
3. When their app incorporates our SDK they get lots of warning messages at built time (an example is in the Actual Result). We reckon it has something to do with the flutter app importing Sentry as well.
4. If they call our SDK which initialises Sentry the app crashes. The app and the SDK have unrelated Sentry DSNs.

We've created two much simpler github repos to help illustrate and reproduce the problems:

1. https://github.com/Captur/FlutterAppForSentryTesting  which uses the following swift package
2. https://github.com/Captur/SwiftPackageForSentry which incorporates Sentry for swift.

The flutter repo includes:
```
  sentry_dio: 7.6.3
  sentry_flutter: 7.6.3
```
in `pubspec.yaml` (note: we tried using 7.9.0 but encountered build errors related to Podfile.lock).

The flutter doesn't initialise Sentry directly but could/should do so.

### Steps to reproduce:

1. Configure both Sentry DSNs to ones you have access to.
2. Update the iOS project in the flutter codebase to use an Apple Development Team you have access to.
3. Try to build and run the flutter app e.g. using `flutter run` from the command line. 

Optionally add the import statement to AppDelegate.swift to import the Swift package from the second repo and then add a call to the EntryPoint. These changes will lead the app to crash at startup.

Here's a diff as an example of this optional extra step:
```
diff --git a/ios/Runner/AppDelegate.swift b/ios/Runner/AppDelegate.swift
index 70693e4..8516fa0 100644
--- a/ios/Runner/AppDelegate.swift
+++ b/ios/Runner/AppDelegate.swift
@@ -1,8 +1,12 @@
 import UIKit
 import Flutter
+import SwiftPackageForSentry
 
 @UIApplicationMain
 @objc class AppDelegate: FlutterAppDelegate {
+
+    // var entrypoint = EntryPoint.shared
+    
   override func application(
     _ application: UIApplication,
     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
```

### Expected Result

The app should be able to run correctly and report errors (etc) to Sentry using the DSN configured by the app developers. Similarly the SDK should be able to run correctly and report errors/events/etc. from the SDK to its Sentry DSN.

Ideally, a single coherent set of Sentry's classes would be bound into the app's binary (to stop code bloat); as an alternative the app would have an independent set of Sentry classes that it would use and the SDK would have an independent set of Sentry classes that it uses. 

### Actual Result

```
% flutter run                                                                          
Launching lib/main.dart on Julian's iPhone in debug mode...
Automatically signing iOS for device deployment using specified development team in Xcode project: <removed>
Running Xcode build...                                                  
 └─Compiling, linking and signing...                         4.8s
Xcode build done.                                           17.3s
(lldb) objc[831]: Class PrivateSentrySDKOnly is implemented in both /private/var/containers/Bundle/Application/AC0FB31B-FD56-46C5-B5B0-FC268054165E/Runner.app/Frameworks/Sentry.framework/Sentry (0x105045e20) and /private/var/containers/Bundle/Application/AC0FB31B-FD56-46C5-B5B0-FC268054165E/Runner.app/Runner (0x1046e41b0). One of the two will be used. Which one is undefined.
... lots of lines removed from here. The full set is in the attached file.
[flutter-and-swift-both-using-sentry.log](https://github.com/getsentry/sentry-dart/files/12469461/flutter-and-swift-both-using-sentry.log)

objc[831]: Class _TtC13SentryPrivate15SwiftDescriptor is implemented in both /private/var/containers/Bundle/Application/AC0FB31B-FD56-46C5-B5B0-FC268054165E/Runner.app/Frameworks/SentryPrivate.framework/SentryPrivate (0x104a39b78) and /private/var/containers/Bundle/Application/AC0FB31B-FD56-46C5-B5B0-FC268054165E/Runner.app/Runner (0x1046e8468). One of the two will be used. Which one is undefined.
[VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(37)] Using the Impeller rendering backend.
Installing and launching...                                        23.3s
```

The crash that occurs when the SDK's entrypoint is called is:
```
-[SentryOptions enableTimeToFullDisplay]: unrecognized selector sent to instance 0x143f0bcb0
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[SentryOptions enableTimeToFullDisplay]: unrecognized selector sent to instance 0x143f0bcb0'
*** First throw call stack:
(0x1bfe60cb4 0x1b8f003d0 0x1bffd5ab8 0x1bfe770e8 0x1bfedd900 0x103807dac 0x102dbd850 0x102dbbe18 0x102dbbf1c 0x102d42938 0x102d4289c 0x102d427f0 0x102d427b0 0x1c7327eac 0x1c73296ec 0x102d4284c 0x102d42058 0x102d420e4 0x1c2c16330 0x1c2262ba8 0x102d421ac 0x1df408dec)
libc++abi: terminating due to uncaught exception of type NSException
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00000001fee1d578 libsystem_kernel.dylib`__pthread_kill + 8
libsystem_kernel.dylib`:
->  0x1fee1d578 <+8>:  b.lo   0x1fee1d598               ; <+40>
    0x1fee1d57c <+12>: pacibsp 
    0x1fee1d580 <+16>: stp    x29, x30, [sp, #-0x10]!
    0x1fee1d584 <+20>: mov    x29, sp
Target 0: (Runner) stopped.
Installing and launching...                                        24.5s
Error launching application on Julian's iPhone.

```

### Are you willing to submit a PR?

None

**Comment by julianharty** [comment]
BTW I'm willing to submit a PR (that info seems to have disappeared on submission) and I've no idea what Obfuscation and Debug Info are set to. If you can explain what to check I'd be happy to add the info to this ticket.

**Comment by julianharty** [comment]
The attachment with all the duplicate classes info also seems to have gone astray on submission, here it is.
[flutter-and-swift-both-using-sentry.log](https://github.com/getsentry/sentry-dart/files/12469485/flutter-and-swift-both-using-sentry.log)

**Comment by julianharty** [comment]
I'm curious whether using a specific/custom namespace for your iOS/Swift classes here would be enough to disambiguate the conflict - what do you think? it'd still mean there'd be 2 instances of each class presumably but could enable both the app and the SDK to use Sentry successfully and unambiguously.

**Comment by vaind** [comment]
AFAIK the use case you're describing is currently not officially supported - Sentry SDK shouldn't be used in a library that will be included in other libraries/apps which in turn may want to use Sentry themselves. Besides the symbol conflicts you're getting, there would be issues with signal handlers and deciding which error belongs to which instance (your library vs the app) - it's not a simple case of checking the stack trace.

See the note at the top here: https://docs.sentry.io/platforms/ and possibly this issue: https://github.com/getsentry/sentry-docs/issues/3611

**Comment by julianharty** [comment]
@vaind Thank you so much for your speedy, clear, and helpful comment. FYI we started using Sentry in our SDKs about 9 months ago - predating these tickets and decisions and I hadn't spotted the recent clarifications in https://docs.sentry.io/platforms/ , https://github.com/getsentry/sentry-docs/issues/3611 and the related tickets. I've enjoyed reading about the design of your SDKs and message transport, etc. 

We're doing some experiments in a couple of directions - a) to see if we can disambiguate the Sentry instances (this may not go very far if we end up swimming against the tide of how the codebase is designed and implemented, and b) to use another SDK provider and service where any of our customers already use Sentry. Director b) is more expedient and the outcome more predictable so we're following that for this particular case I explained in this ticket.

FWIW I've worked with various mobile analytics SDKs and related services and am continuing to learn on an ongoing basis of various complexities that need to be addressed to use mobile analytics well. Thank you again.

I'm happy to close this ticket - let me know if there's any purpose in keeping it open for longer.
